keps/sig-cli/3659-kubectl-apply-prune/README.md

```diff
@@ -0,0 +1,1365 @@
+- An ApplySet object can be part of another ApplySet (sub-ApplySets).
+
+### ApplySet Naming
+
+Each ApplySet MUST have an ID that can be used to uniquely identify the parent and member objects via the label selector conventions outlined in the following sections. As such, the name:
+* is subject to the normal limits of label values
+* MUST be generated in a way that provides reasonable uniqueness guarantees at the cluster scope
+
+This uniqueness constraint is designed to prevent over-selection due to naming collisions within overlapping scopes. To comply with it, this specification recommends that tooling choose IDs derived from identifiers that are already globally unique. For instance, the parent object's UUID or a base64 encoding of its GKNN would be suitable. We reserve the right to dictate a specific format at some point before this KEP goes into beta.
+
+This name does not need to be used for the `metadata.name` of the parent object. In fact, it is likely desirable for tooling to allow end users to choose the `metadata.name` of the parent so that it is more intuitive for them to refer to.
+
+### ApplySet Member Objects
+
+#### Labels
+
+Objects that are part of an ApplySet MUST carry the following label:
+
+```yaml
+applyset.k8s.io/part-of: <applysetkey> # REQUIRED
+```
+
+This `applyset.k8s.io/part-of` label is the source of truth for membership in a set. Its `<applysetkey>` value MUST match the value of `applyset.k8s.io/id` on the parent (see [ApplySet Parent Objects](#labels-and-annotations)) and comply with the naming constraints outlined in [ApplySet Naming](#applyset-naming).
+
+
+### ApplySet Parent Objects
+
+Although ApplySet parent objects can (in theory) be of any type, for both performance and simplicity, we choose to limit supported types to the following:
+- ConfigMap
+- Secret
+- custom resources, where the CRD registering them is labeled with `applyset.k8s.io/role/parent` (The value is currently ignored, but implementors should set an empty value to be forwards-compatible with future evolution of this convention.)
+
+This list may be extended in the future as use cases are presented. Keeping it minimal has significant benefits for the performance of any commands that list the parent objects themselves. Specific tools may further limit the types they support, but any tooling designed to identify all ApplySets must consider the exact list set out above.
+
+While a purpose-made cluster-scoped CRD is a logical choice, the specification has no opinion on this, so as to accommodate the many and various choices existing tools have already made. In choosing types to support as parents, tools should consider what permissions their target users typically have on their clusters; for instance, they may not have permissions to install CRDs.
+
+#### Labels and annotations
+
+ApplySets MUST also have a “parent object” in the cluster. The ApplySet object MUST be labeled with:
+
+```yaml
+applyset.k8s.io/id: <applysetkey> # REQUIRED
```

Comment:
what happens if two different objects in the cluster are labeled with the same applysetkey?

Reply:
Bad things (tm).  We have two suggested strategies that we want to explore: base64 of GKNN and a random GUID.  I suspect we'll start with the random GUID.

Reply:
Actually, I think you've persuaded me on chat that we should start with the encoded GKNN/GVKNN, because of the "confused parent attack"

Reply:
We have now made this change - thank you for the suggestion!